README.md
# logseq-plugin-git
> git plugin for logseq

[![latest release version](https://img.shields.io/github/v/release/haydenull/logseq-plugin-git)](https://github.com/haydenull/logseq-plugin-git/releases)
[![License](https://img.shields.io/github/license/haydenull/logseq-plugin-git?color=blue)](https://github.com/haydenull/logseq-plugin-git/blob/main/LICENSE)

Logseq Git Plugin is a plugin for Logseq that provides easy access to common git commands, and helps you keep your notes synchronized with a remote git repository.

With this plugin, you can:

- Know explicitly if there are unsaved notes in the current repository and if they are synchronized with the remote repository. When there are unsaved files locally, the plugin will show a red indicator in the toolbar icon.
- Quickly commit changes to notes. The plugin provides shortcut keys (mod+s) to perform commit&push operations.
- Provide buttons to quickly execute common commands. Click the plugin icon, and the drop-down menu will show you some buttons to help you quickly execute commands (you can specify the desired buttons in the settings).
- Prompt you if you have synchronized with the remote repository. The plugin automatically detects if the current repository is up-to-date and prompts when logseq starts and other times.

## Installation

Install from the Logseq Plugin Store

## Shortcuts
The plugin provides a shortcut key (mod+s) to quickly commit and push changes to the remote repository.

## Toolbar Icon

- icon: everything is up-to-date.
- icon with red dot: there are unsaved files locally.
- icon with blinking red dot: the plugin is committing or pushing changes to the remote repository.

## Commands

The drop-down menu shows the following commands:
- Check Status: check the status of the current repository.
- Show Log: show the log of the current repository.
- Pull: pull changes from the remote repository.
- Pull Rebase: pull changes from the remote repository and rebase.
- Commit: commit changes to the current repository.
- Push: push changes to the remote repository.
- Commit & Push: commit changes and push to the remote repository.

## Demo
![demo](./demo.png)



CHANGELOG.md
## [1.5.5](https://github.com/haydenull/logseq-plugin-git/compare/v1.5.4...v1.5.5) (2023-12-28)


### Bug Fixes

* patch git exec command ([#50](https://github.com/haydenull/logseq-plugin-git/issues/50)) ([b4a4229](https://github.com/haydenull/logseq-plugin-git/commit/b4a422930288ff32cf2ff2fd23a71f0ca862b606))

## [1.5.4](https://github.com/haydenull/logseq-plugin-git/compare/v1.5.3...v1.5.4) (2023-07-27)


### Bug Fixes

* exported inProcess() was not set correctly breaking checkIsSynced ([#37](https://github.com/haydenull/logseq-plugin-git/issues/37)) ([cf78518](https://github.com/haydenull/logseq-plugin-git/commit/cf785180e4f68e589a153b71056a2fe1163fade6))

## [1.5.3](https://github.com/haydenull/logseq-plugin-git/compare/v1.5.2...v1.5.3) (2023-07-26)


### Bug Fixes

* [#30](https://github.com/haydenull/logseq-plugin-git/issues/30), shortcut mod+alt+s will pull rebase ([5f42bd5](https://github.com/haydenull/logseq-plugin-git/commit/5f42bd5674cdcdff6ae60e789ddf9bc66223ed38))
* stack git operations not to get errors when auto actions are already performing ([e78bb2c](https://github.com/haydenull/logseq-plugin-git/commit/e78bb2c185787e731cb7b1af07487b6901389396)), closes [#34](https://github.com/haydenull/logseq-plugin-git/issues/34)

## [1.5.2](https://github.com/haydenull/logseq-plugin-git/compare/v1.5.1...v1.5.2) (2023-04-20)


### Bug Fixes

* :bug: the mask was not turned off after initilization ([a84acb5](https://github.com/haydenull/logseq-plugin-git/commit/a84acb5a7133da1f0693c15112ab5c903f0a1774))

## [1.5.1](https://github.com/haydenull/logseq-plugin-git/compare/v1.5.0...v1.5.1) (2023-04-16)


### Bug Fixes

* In the Logseq 0.9.2 version, there is an issue with the panel display that needs to be fixed. ([a81bceb](https://github.com/haydenull/logseq-plugin-git/commit/a81bceb10996a6e3d3901fe1f1fef63d6d1c4052))

# [1.5.0](https://github.com/haydenull/logseq-plugin-git/compare/v1.4.2...v1.5.0) (2023-02-14)


### Features

* logseq sdk version ([#26](https://github.com/haydenull/logseq-plugin-git/issues/26)) ([d616d82](https://github.com/haydenull/logseq-plugin-git/commit/d616d820df800809cdc37e718fa0a43ff380498b))

## [1.4.2](https://github.com/haydenull/logseq-plugin-git/compare/v1.4.1...v1.4.2) (2022-11-06)


### Bug Fixes

* button color ([d3a5bd1](https://github.com/haydenull/logseq-plugin-git/commit/d3a5bd169ee74634983c04f160677a9f34f8971e))

## [1.4.1](https://github.com/haydenull/logseq-plugin-git/compare/v1.4.0...v1.4.1) (2022-11-01)


### Bug Fixes

* button style ([d32ef80](https://github.com/haydenull/logseq-plugin-git/commit/d32ef80c2105b3d7fb14da9b813c2eb29b7d3045))

# [1.4.0](https://github.com/haydenull/logseq-plugin-git/compare/v1.3.0...v1.4.0) (2022-08-23)


### Features

* ✨ Support for turning off auto-check sync ([a2ac2bd](https://github.com/haydenull/logseq-plugin-git/commit/a2ac2bd44cd9685c15827bdb82bdec6c18139823))

# [1.3.0](https://github.com/haydenull/logseq-plugin-git/compare/v1.2.0...v1.3.0) (2022-08-19)


### Features

* add default value: pull rebase ([90862ba](https://github.com/haydenull/logseq-plugin-git/commit/90862ba1d2d0d258c6d7b7a58f552fb0187f2df9))

# [1.2.0](https://github.com/haydenull/logseq-plugin-git/compare/v1.1.0...v1.2.0) (2022-06-11)


### Features

* supports disable checkWhenDBChanged ([85428aa](https://github.com/haydenull/logseq-plugin-git/commit/85428aa217361a797f64c5eadd8a008e39cb49b2))

# [1.1.0](https://github.com/haydenull/logseq-plugin-git/compare/v1.0.2...v1.1.0) (2022-05-22)


### Features

* ✨ add pull and checkout buttons ([f950626](https://github.com/haydenull/logseq-plugin-git/commit/f950626e2a3776c805a96267bbd37ba7a7eae9da))

## [1.0.2](https://github.com/haydenull/logseq-plugin-git/compare/v1.0.1...v1.0.2) (2022-05-20)


### Bug Fixes

* The id does not match expectations ([dffcaa3](https://github.com/haydenull/logseq-plugin-git/commit/dffcaa3b71086bad022350494c841bed7576d9c3))

## [1.0.1](https://github.com/haydenull/logseq-plugin-git/compare/v1.0.0...v1.0.1) (2022-05-05)


### Bug Fixes

* 🐛 Adapt logseq 0.6.7 ([02ffdf7](https://github.com/haydenull/logseq-plugin-git/commit/02ffdf70d493f0adc70d95d4847a7bb6ed6751b7))
* default value error[skip ci] ([a5cfc2b](https://github.com/haydenull/logseq-plugin-git/commit/a5cfc2b9184119f820946fcb0d33f5e5dc098e5d))

# 1.0.0 (2022-04-08)


### Features

* ✨ add icon ([0f52416](https://github.com/haydenull/logseq-plugin-git/commit/0f52416ef8594525fb2fb527bc05c98fa327e308))

# [1.1.0](https://github.com/haydenull/logseq-plugin-git/compare/v1.0.0...v1.1.0) (2022-03-08)


### Features

* add antd ([036577d](https://github.com/haydenull/logseq-plugin-git/commit/036577dc529db4e4a5964c287a55d112bae654bc))

# 1.0.0 (2022-01-18)


### Features

* ✨ initial commit ([cca0e7f](https://github.com/haydenull/logseq-plugin-git/commit/cca0e7fcba33830eaf534fd9ca6b867b57147de4))


src/helper/git.ts
// https://logseq.github.io/plugins/interfaces/IAppProxy.html#execGitCommand
import type { IGitResult } from "@logseq/libs/dist/LSPlugin.user"

export const status = async (showRes = true): Promise<IGitResult> => {
  // git status --porcelain | awk '{print $2}'
  // git status --porcelain | wc -l
  const res =  await logseq.Git.execCommand(['status', '--porcelain'])
  console.log('[faiz:] === git status', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git status success')
    } else {
      logseq.UI.showMsg(`Git status failed\n${res.stderr}`, 'error')
    }
  }
  /**
   * res
   * modify
   * {
   *  exitCode: 0,
   *  stderr: '',
   *  stdout: 'M foo.md\n?? bar.md\n',
   * }
   * ahead & uptodate & behind
   * {
   * exitCode: 0,
   * stderr: '',
   * stdout: '',
   * }
   */
  // changed files    staged files
  return res
}

// log with git log --pretty=format:"%h %ad | %s%d [%an]" --date=short
export const log = async (showRes = true): Promise<IGitResult> => {
  // git log --pretty=format:"%h %s" -n 1
  // git log --pretty=format:"%h %ad | %s%d [%an]" --date=short
  // return await logseq.App.execGitCommand(['log', '--pretty=format:"%h %s"'])
  const res = await logseq.Git.execCommand(['log', '--pretty=format:"%h %ad | %s [%an]"', '--date=format:"%Y-%m-%d %H:%M:%S"', '--name-status'])
  console.log('[faiz:] === git log', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git log success')
    } else {
      logseq.UI.showMsg(`Git log failed\n${res.stderr}`, 'error')
    }
  }
  return res
}

// git pull
export const pull = async (showRes = true): Promise<IGitResult> => {
  const res = await logseq.Git.execCommand(['pull'])
  console.log('[faiz:] === git pull', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git pull success')
    } else {
      logseq.UI.showMsg(`Git pull failed\n${res.stderr}`, 'error')
    }
  }
  return res
}

// git pull --rebase
export const pullRebase = async (showRes = true): Promise<IGitResult> => {
  const res = await logseq.Git.execCommand(['pull', '--rebase'])
  console.log('[faiz:] === git pull --rebase', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git pull --rebase success')
    } else {
      logseq.UI.showMsg(`Git pull --rebase failed\n${res.stderr}`, 'error')
    }
  }
  return res
}

// git checkout .
export const checkout = async (showRes = true): Promise<IGitResult> => {
  const res = await logseq.Git.execCommand(['checkout', '.'])
  console.log('[faiz:] === git checkout .', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git checkout success')
    } else {
      logseq.UI.showMsg(`Git checkout failed\n${res.stderr}`, 'error')
    }
  }
  return res
}

// git commit
export const commit = async (showRes = true, message: string): Promise<IGitResult> => {
  await logseq.Git.execCommand(['add', '.'])
  // git commit -m "message"
  const res = await logseq.Git.execCommand(['commit', '-m', message])
  console.log('[faiz:] === git commit', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git commit success')
    } else {
      logseq.UI.showMsg(`Git commit failed\n${res.stdout || res.stderr}`, 'error')
    }
  }
  return res
}

// push
export const push = async (showRes = true): Promise<IGitResult> => {
  // git push
  const res = await logseq.Git.execCommand(['push'])
  console.log('[faiz:] === git push', res)
  if (showRes) {
    if (res.exitCode === 0) {
      logseq.UI.showMsg('Git push success')
    } else {
      logseq.UI.showMsg(`Git push failed\n${res.stderr}`, 'error')
    }
  }
  return res
}

src/helper/util.ts
import {
  ACTIVE_STYLE,
  HIDE_POPUP_STYLE,
  INACTIVE_STYLE,
  SHOW_POPUP_STYLE,
} from "./constants";
import { status, inProgress, execGitCommand } from "./git";

export const checkStatus = async () => {
  console.log("Checking status...");
  const statusRes = await status(false);
  if (statusRes?.stdout === "") {
    console.log("No changes", statusRes);
    setPluginStyle(INACTIVE_STYLE);
  } else {
    console.log("Need save", statusRes);
    setPluginStyle(ACTIVE_STYLE);
  }
  return statusRes;
};

let pluginStyle = "";
export const setPluginStyle = (style: string) => {
  pluginStyle = style;
  logseq.provideStyle({ key: "git", style });
};
export const getPluginStyle = () => pluginStyle;

export const showPopup = () => {
  const _style = getPluginStyle();
  logseq.UI.queryElementRect("#logseq-git--git").then((triggerIconRect) => {
    console.log("[faiz:] === triggerIconRect", triggerIconRect);
    if (!triggerIconRect) return;
    const popupWidth = 120 + 10 * 2;
    const left =
      triggerIconRect.left + triggerIconRect.width / 2 - popupWidth / 2;
    const top = triggerIconRect.top + triggerIconRect.height;
    const _style = getPluginStyle();
    setPluginStyle(
      `${_style}\n.plugin-git-popup{left:${left}px;top:${top}px;}`
    );
  });
  setPluginStyle(`${_style}\n${SHOW_POPUP_STYLE}`);
};
export const hidePopup = () => {
  const _style = getPluginStyle();
  setPluginStyle(`${_style}\n${HIDE_POPUP_STYLE}`);
};

export const debounce = (fn, wait: number = 100, environment?: any) => {
  let timer = null;
  return function () {
    // @ts-ignore
    const context = environment || this;
    const args = arguments;
    if (timer) {
      clearTimeout(timer);
      timer = null;
    }
    // @ts-ignore
    timer = setTimeout(function () {
      fn.apply(context, args);
    }, wait);
  };
};

export const checkStatusWithDebounce = debounce(() => {
  checkStatus();
}, 2000);

export const isRepoUpTodate = async () => {
  await execGitCommand(["fetch"]);
  const local = await execGitCommand(["rev-parse", "HEAD"]);
  const remote = await execGitCommand(["rev-parse", "@{u}"]);
  logseq.UI.showMsg(`${local.stdout} === ${remote.stdout}`, "success", { timeout: 30 });
  return local.stdout === remote.stdout;
};

export const checkIsSynced = async () => {
  if (inProgress()) {
    console.log("[faiz:] === checkIsSynced Git in progress, skip check");
    return
  }

  const isSynced = await isRepoUpTodate();
  if (!isSynced)
    logseq.UI.showMsg(
      `The current repository is not synchronized with the remote repository, please check.`,
      "warning",
      { timeout: 0 }
    );
};


src/helper/styles.ts
import { ACTIVE_STYLE, GIT_STATUS, INACTIVE_STYLE } from "./constants";
import { debounce, logDebug, logWarn } from "./util";

let gitStatus = 0;
let loadingCount = 0;
let popupLeft = 0;
let popupTop = 0;

const COMMON_STYLE = `
.ui-items-container[data-type=toolbar] > .list-wrap {
  overflow: visible;
}
#injected-ui-item-git-logseq-git {
  position: relative;
}
.plugin-git-container {
  display: none;
}
.plugin-git-container .plugin-git-mask {
  position: fixed;
  width: 100vw;
  height: 100vh;
  left: 0;
  top: 0;
  z-index: 99;
}
.plugin-git-container .plugin-git-popup {
  position: fixed;
  z-index: 99;
  background-color: var(--ls-secondary-background-color);
  padding: 10px;
  border-radius: .375rem;
  --tw-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}
.plugin-git-container .plugin-git-popup::before {
  content: '';
  position: absolute;
  top: -8px;
  left: 50%;
  transform: translateX(-50%);
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 0 10px 8px 10px;
  border-color: transparent transparent var(--ls-secondary-background-color) transparent;
}
`;

const DISABLED_STYLE = `
#injected-ui-item-git-logseq-git::after {
  display: none;
}
#injected-ui-item-git-logseq-git .ti-brand-git:before {
  color: var(--color-level-4);
}`;

const REMOTE_OK_STYLE = `#injected-ui-item-git-logseq-git::before {
  display: block;
  content: '';
  position: absolute;
  width: 8px;
  height: 8px;
  border-radius: 100%;
  background-color: var(--ls-color-file-sync-idle);
  left: 4px;
  top: 6px;
}`;
const REMOTE_CHANGED_OR_ERROR_STYLE = `#injected-ui-item-git-logseq-git::before {
  display: block;
  content: '';
  position: absolute;
  width: 8px;
  height: 8px;
  border-radius: 100%;
  background-color: var(--ls-color-file-sync-pending);
  left: 4px;
  top: 6px;
}`;

const LOCAL_OK_STYLE = `#injected-ui-item-git-logseq-git::after {
  display: block;
  content: '';
  position: absolute;
  width: 8px;
  height: 8px;
  border-radius: 100%;
  background-color: var(--ls-color-file-sync-idle);
  right: 4px;
  top: 6px;
}
`;
const LOCAL_CHANGED_OR_ERROR_STYLE = `
#injected-ui-item-git-logseq-git::after {
  display: block;
  content: '';
  position: absolute;
  width: 8px;
  height: 8px;
  border-radius: 100%;
  background-color: var(--ls-color-file-sync-pending);
  right: 4px;
  top: 6px;
}
`;

export const SHOW_POPUP_STYLE = () => `
.plugin-git-container {
  display: block;
}
.plugin-git-popup{
  left:${popupLeft}px;
  top:${popupTop}px;
}
`;

export const LOADING_STYLE = `
#injected-ui-item-git-logseq-git .ti-brand-git:before {
  animation: blinker 1s linear infinite;
}
@keyframes blinker {
  50% {
    opacity: 0;
  }
}
`;

export const setGitStateChange = (statusChange: GIT_STATUS) => {
  switch (statusChange) {
    case GIT_STATUS.Disabled:
      gitStatus = 0;
      break;
    case GIT_STATUS.LocalInUse:
    case GIT_STATUS.RemoteInUse:
      gitStatus = gitStatus | statusChange;
      break;
    case GIT_STATUS.LocalChanges:
      gitStatus = gitStatus | (statusChange & ~GIT_STATUS.LocalError);
      break;
    case GIT_STATUS.LocalOk:
      gitStatus = gitStatus & ~GIT_STATUS.LocalChanges & ~GIT_STATUS.LocalError;
      break;
    case GIT_STATUS.LocalError:
      gitStatus = gitStatus | (statusChange & ~GIT_STATUS.LocalChanges);
      break;
    case GIT_STATUS.RemoteChanges:
      gitStatus = gitStatus | (statusChange & ~GIT_STATUS.RemoteError);
      break;
    case GIT_STATUS.RemoteOk:
      gitStatus =
        gitStatus & ~GIT_STATUS.RemoteChanges & ~GIT_STATUS.RemoteError;
      break;
    case GIT_STATUS.RemoteError:
      gitStatus = gitStatus | (statusChange & ~GIT_STATUS.RemoteChanges);
      break;
  }

  logDebug("setGitStateChange =", GIT_STATUS[statusChange], gitStatus);
  setStatusToStyle();
};

const setStatusToStyle = () => {
  if ((gitStatus & GIT_STATUS.LocalInUse) === 0) {
    _setPluginStyle(DISABLED_STYLE);
    return;
  }

  let localStyle = LOCAL_OK_STYLE;
  if (
    gitStatus & GIT_STATUS.LocalChanges ||
    gitStatus & GIT_STATUS.LocalError
  ) {
    // TODO add style for error state
    localStyle = LOCAL_CHANGED_OR_ERROR_STYLE;
  }

  let remoteStyle = "";
  if (gitStatus & GIT_STATUS.RemoteInUse) {
    if (
      gitStatus & GIT_STATUS.RemoteChanges ||
      gitStatus & GIT_STATUS.RemoteError
    ) {
      // TODO add style for error state
      remoteStyle = REMOTE_CHANGED_OR_ERROR_STYLE;
    } else {
      remoteStyle = REMOTE_OK_STYLE;
    }
  }

  _setPluginStyle(`${localStyle}\n${remoteStyle}`);
};

export const _setPluginStyle = (style: string) => {
  const addStyle: string[] = [];

  if (gitStatus & GIT_STATUS.MenuOpen) {
    addStyle.push(SHOW_POPUP_STYLE());
  }

  if (gitStatus & GIT_STATUS.Loading) {
    addStyle.push(LOADING_STYLE);
  }

  logseq.provideStyle({
    key: "git",
    style: `${COMMON_STYLE}\n${style}\n${addStyle.join("\n")}`
  });
};

const loadingStart = () => {
  loadingCount += 1;
  if (loadingCount === 1) {
    logDebug("loading = START");
    gitStatus = gitStatus | GIT_STATUS.Loading;
    setStatusToStyle();
  }
};

const loadingEnd = () => {
  loadingCount -= 1;
  if (loadingCount === 0) {
    logDebug("loading = END");
    gitStatus = gitStatus & ~GIT_STATUS.Loading;
    setStatusToStyle();
  }
};
export const loadingEffect = async fn => {
  let ret;
  loadingStart();
  try {
    ret = await fn();
  } finally {
    loadingEnd();
  }
  return ret;
};

export const showPopup = () => {
  logDebug("showPopup");
  logseq.UI.queryElementRect("#logseq-git--git").then(triggerIconRect => {
    // logDebug("triggerIconRect", triggerIconRect);
    if (!triggerIconRect) return;
    const popupWidth = 120 + 10 * 2;
    popupLeft =
      triggerIconRect.left + triggerIconRect.width / 2 - popupWidth / 2;
    popupTop = triggerIconRect.top + triggerIconRect.height;
    gitStatus = gitStatus | GIT_STATUS.MenuOpen;
    setStatusToStyle();
  });
};
export const hidePopup = () => {
  logDebug("hidePopup");
  gitStatus = gitStatus & ~GIT_STATUS.MenuOpen;
  setStatusToStyle();
};


src/helper/constants.ts
import { SettingSchemaDesc } from "@logseq/libs/dist/LSPlugin.user";

export const COMMON_STYLE = `
.ui-items-container[data-type=toolbar] > .list-wrap {
  overflow: visible;
}
#injected-ui-item-git-logseq-git {
  position: relative;
}
.plugin-git-container {
  display: none;
}
.plugin-git-container .plugin-git-mask {
  position: fixed;
  width: 100vw;
  height: 100vh;
  left: 0;
  top: 0;
  z-index: 99;
}
.plugin-git-container .plugin-git-popup {
  position: fixed;
  z-index: 99;
  background-color: var(--ls-secondary-background-color);
  padding: 10px;
  border-radius: .375rem;
  --tw-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}
.plugin-git-container .plugin-git-popup::before {
  content: '';
  position: absolute;
  top: -8px;
  left: 50%;
  transform: translateX(-50%);
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 0 10px 8px 10px;
  border-color: transparent transparent var(--ls-secondary-background-color) transparent;
}
`;

export const SHOW_POPUP_STYLE = `
.plugin-git-container {
  display: block;
}
`;
export const HIDE_POPUP_STYLE = `
.plugin-git-container {
  display: none;
}
`;

export const INACTIVE_STYLE = `
${COMMON_STYLE}
#injected-ui-item-git-logseq-git::after {
  display: none;
}
`;
export const ACTIVE_STYLE = `
${COMMON_STYLE}
#injected-ui-item-git-logseq-git::after {
  display: block;
  content: '';
  position: absolute;
  width: 8px;
  height: 8px;
  border-radius: 100%;
  background-color: rgb(237, 66, 69);
  right: 8px;
  top: 6px;
}
`;

export const LOADING_STYLE = `
${COMMON_STYLE}
#injected-ui-item-git-logseq-git::after {
  display: block;
  content: '';
  position: absolute;
  width: 8px;
  height: 8px;
  border-radius: 100%;
  background-color: rgb(237, 66, 69);
  right: 8px;
  top: 6px;
  animation: blink 1s linear infinite;
}
@keyframes blink {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}
`;

export const BUTTONS = [
  { key: "status", title: "Check Status", event: "check" },
  { key: "log", title: "Show Log", event: "log" },
  { key: "pull", title: "Pull", event: "pull" },
  { key: "pullRebase", title: "Pull Rebase", event: "pullRebase" },
  { key: "checkout", title: "Checkout", event: "checkout" },
  { key: "commit", title: "Commit", event: "commit" },
  { key: "push", title: "Push", event: "push" },
  { key: "commitAndPush", title: "Commit & Push", event: "commitAndPush" },
];

export const SETTINGS_SCHEMA: SettingSchemaDesc[] = [
  {
    key: "buttons",
    title: "Buttons",
    type: "enum",
    default: ["Check Status", "Show Log", "Pull Rebase", "Commit & Push"],
    description: "Select buttons to show",
    enumPicker: "checkbox",
    enumChoices: BUTTONS.map(({ title }) => title),
  },
  {
    key: "checkWhenDBChanged",
    title: "Check Status when DB Changed",
    type: "boolean",
    default: true,
    description: "Check status when DB changed, restart logseq to take effect",
  },
  {
    key: "autoCheckSynced",
    title: "Auto Check If Synced",
    type: "boolean",
    default: false,
    description:
      "Automatically check if the local version is the same as the remote",
  },
  {
    key: "autoPush",
    title: "Auto Push",
    type: "boolean",
    default: false,
    description: "Auto push when logseq hide",
  },
];
